docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,430 @@
+```
+
+The wrapper implementation of `IDemoGetType.GetString()` should consist of marshalling logic and then a dispatch to the managed object being wrapped. All the state for dispatch is contained within the provided `_this` argument. The `_this` argument will actually be of type `ComInterfaceDispatch*`. This type represents a low-level structure with a single field, `Vtable`, that will be discussed later. Further details of this type and its layout are an implementation detail of the runtime and should not be depended upon. In order to retrieve the managed instance from a `ComInterfaceDispatch*` instance, use the following code:
+
+```csharp
+IDemoGetType inst = ComInterfaceDispatch.GetInstance<IDemoGetType>((ComInterfaceDispatch*)_this);
+```
+
+Now that you have a C# method that can be inserted into a vtable, you can construct the vtable. Note the use of [`RuntimeHelpers.AllocateTypeAssociatedMemory()`][api_allocatetypeassociatedmemory] for allocating memory in a way that works with [unloadable][doc_unloadability] assemblies.
+
+```csharp
+GetIUnknownImpl(
+    out IntPtr fpQueryInterface,
+    out IntPtr fpAddRef,
+    out IntPtr fpRelease);
+
+{
+    int tableCount = 4;
+    int idx = 0;
+    var vtable = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(
+        typeof(DemoComWrappers),
+        IntPtr.Size * tableCount);
+    vtable[idx++] = fpQueryInterface;
+    vtable[idx++] = fpAddRef;
+    vtable[idx++] = fpRelease;
+    vtable[idx++] = (IntPtr)(delegate* unmanaged<IntPtr, IntPtr*, int>)&ABI.
+IDemoGetTypeManagedWrapper.GetString;
+    Debug.Assert(tableCount == idx);
+    s_IDemoGetTypeVTable = (IntPtr)vtable;
+}
+```
+
+The allocation of vtables is the first part of implementing `ComputeVtables()`. You should also construct comprehensive definitions for types that you're planning to support &ndash; think `DemoImpl`. Using the constructed vtables, you can now create a series of [`ComInterfaceEntry`][api_cominterfaceentry] instances that represent the complete view of the managed object in COM.
+
+```csharp
+{
+    s_DemoImplDefinitionLen = 2;
+    int idx = 0;
+    var entries = (ComInterfaceEntry*)RuntimeHelpers.AllocateTypeAssociatedMemory(
+        typeof(DemoComWrappers),
+        sizeof(ComInterfaceEntry) * s_DemoImplDefinitionLen);
+    entries[idx].IID = IDemoGetType.IID_IDemoGetType;
+    entries[idx++].Vtable = s_IDemoGetTypeVTable;
+    entries[idx].IID = IDemoStoreType.IID_IDemoStoreType;
+    entries[idx++].Vtable = s_IDemoStoreVTable;
+    Debug.Assert(s_DemoImplDefinitionLen == idx);
+    s_DemoImplDefinition = entries;
+}
+```
+
+The allocation of vtables and entries for the Managed Object Wrapper can and should be done ahead of time since the data can be used for all instances of the type. The work here could be performed in a `static` constructor or a module initializer, but it should be done ahead of time so the `ComputeVtables()` method is as simple and quick as possible.
+
```

Comment:
That is fair. In this case those I would argue the types the ComWrappers implementation supports should be left alone because you never know what type will appear from COM.

Reply:
> the types the ComWrappers implementation supports should be left alone because you never know what type will appear from COM.

That is absolutely true - but what if the code which uses the ComWrappers is not used... in that case the entire ComWrappers class can be removed. Unless it's rooted somewhere else, like the module constructor.
It's also at odds with "pay for play" - the app would pay the price of initializing it up front, without knowing if it will need it or not.

Reply:
>It's also at odds with "pay for play" - the app would pay the price of initializing it up front, without knowing if it will need it or not.

I would like to think more about that UX. The `ComWrappers` system is going to include substantial code gen, so why generate it in the first place if it isn't going to be needed? In fact that seems like a layering issue rather than push something so big into the module initializer vs a `static` type initializer. Perhaps a compromise here would be to add a comment that states module initializers may not be the best approach for trimmer support?